### ISSUE TITLE
Train/test scores from QSARtuna model

### ISSUE THREAD

**Issue by jackhhewitt** [MolecularAI/QSARtuna#41]
When running the following code:
```python
build_best(buildconfig_best(study), "../model.pkl")

import pickle
with open("../model.pkl", "rb") as m:
    model = pickle.load(m)

print(model)
``` 
I receive an output containing things like the predictor, descriptors, parameters etc, but I also receive:
```python
train_scores {accuracy: 0.9, average precision: 0.98} etc etc 
``` 
```python
test_scores {accuracy: 0.6, average precision: 0.13} etc etc
``` 

My question is, how are these calculated? Are the training scores calculated during the cross-validation and the test scores from predicting with the provided 'test_dataset_file'? And what do balanced, macro/micro mean?

Much appreciated

**Comment by lewismervin1** [comment]
Hi @jackhhewitt, thanks for the question and welcome!

There is some overlap with your question and the one here: https://github.com/MolecularAI/QSARtuna/issues/39. 

I will also answer here for clarity. The `build_best` method will train a model on the best performing hyper-parameters. It evaluates the `train_score` on the data used to train the model (i.e. full leakage since it has seen all this data before) and evaluates the `test_scores` on the the data reserved from the model (i.e. the outer test split that has zero leakage). 

If you supplied a `test_dataset_file` then your `test_scores` performance will be based on this. Please note that `test_scores`  may also be evaluated on test data from your training data file if you provided a `split_strategy` or did not over-ride the default strategy.



The `macro` and `micro` appendages to scores denote the manner in which the metric is calculated over the different classes in your model, which is a nomenclature used by the scikit-learn package. 

You can see a commentary on their conventions here: https://scikit-learn.org/stable/modules/model_evaluation.html#from-binary-to-multiclass-and-multilabel.

 Generally speaking `macro` simply calculates the mean of the binary metrics, giving equal weight to each class, whereas `micro` gives each sample-class pair an equal contribution to the overall metric. I plan to add more metrics that support `weighted` aggregation options in the future, which will allow us to account for class imbalance by computing the average of binary metrics in the future.

I hope this clarifies things. Please let me know if not and thanks for using QSARtuna!

Lewis

**Comment by jackhhewitt** [comment]
> Hi [@jackhhewitt](https://github.com/jackhhewitt), thanks for the question and welcome!
> 
> There is some overlap with your question and the one here: [#39](https://github.com/MolecularAI/QSARtuna/issues/39).
> 
> I will also answer here for clarity. The `build_best` method will train a model on the best performing hyper-parameters. It evaluates the `train_scores` on the data used to train the model (i.e. full leakage since it has seen all this data before) and evaluates the `test_scores` on the the data reserved from the model (i.e. the outer test split that has zero leakage).
> 
> If you supplied a `test_dataset_file` then your `test_scores` performance will be based on this. Please note that `test_scores` may also be evaluated on a portion of data reserved from your training data file _if_ you provided a `split_strategy` or did not over-ride the default strategy.
> 
> The `macro` and `micro` appendages to scores denote the manner in which the metric is calculated over the different classes in your model, which is a nomenclature used by the scikit-learn package.
> 
> You can see a commentary on their conventions here: https://scikit-learn.org/stable/modules/model_evaluation.html#from-binary-to-multiclass-and-multilabel.
> 
> Generally speaking `macro` simply calculates the mean of the binary metrics, giving equal weight to each class, whereas `micro` gives each sample-class pair an equal contribution to the overall metric. I plan to add more metrics that support `weighted` aggregation options in the future, which will allow us to factor class imbalance between active/inactive sets by computing the average of binary metrics in which each class’s score weighted by its presence in the true data sample.
> 
> I hope this clarifies things. Please let me know if not and thanks for using QSARtuna!
> 
> Lewis

Hi Lewis, 

Thanks for your reply. This clarifies things well. 

One more thing, looking at these 'test_scores', when I use 'model.predict_from_smiles(test_smiles)', returned is a list of numbers that are neither 1 nor 0, am I correct in assuming there is a threshold applied to convert say 0.612 -> 1 and say 0.323 ->0, like a default 0.5? And where can this threshold be set or found?

Many thanks, 
Jack

**Comment by jackhhewitt** [comment]
This can be marked as closed now as I've sorted this :)
